### DISCUSSION TITLE
Reformat code base

### DISCUSSION THREAD

**Discussion by pr8x** [AvaloniaUI/Avalonia#6261] [General]
Are there any plans to reformat the entire code base at some point? There are quite few spots with inconsistent indentation or order of properties/fields/functions.

**Comment by robloo** [comment]
In terms of controls I would expand this topic even further: Refactor Avalonia.Controls #4957

**Reply by grokys** [reply]
As I said, I'm all for consistency but -1 on this suggestion. I spend most of my days in Avalonia.Controls and don't feel that would help me in any meaningful way.

**Reply by grokys** [reply]
Though tbh I think @maxkatz6 spends more of his time in Avalonia.Controls that me, so he might want to chip in.

**Reply by robloo** [reply]
Even doing it one at a time would be a help. All the Grid related classes should be cleaned up to start. Then a new Buttons directory will be needed soon.

**Reply by robloo** [reply]
The current inconsistency in directory also bothers me :) ItemsRepeater for example you decided to put in one directory. That's done for a lot of the newer control additions. Flyout, etc. However the original controls are put all over the place. In huge projects it starts to make sense to organize by feature. Even MVVM monolithic ViewModel folders usually turns into ViewModels being placed in a directly along with the View itself. Its just a better way to organize large projects: by feature, IMO.

**Reply by grokys** [reply]
The problem with putting everything in directories is that VS uses that directory structure for the namespace when creating new files, meaning that people start committing files with incorrect namespaces. I've done it a few times myself but thankfully noticed before submitting a PR. We've got a few examples of that happening still in the codebase and we can't fix the namespaces until the next major release because it's a breaking change.

Because of this, directories that don't match the namespace of the classes within should be avoided unless really necessary (i.e. when a control has a lot of files in it (i.e. ported from WinUI)). Even then, I'm not sure it's the correct choice.

I'd consider changing my position on this only when VS gives us some way to not create broken APIs by default (I think Rider has the same behavior too).

**Reply by maxkatz6** [reply]
Rider handles it better. If files in the folder do not have new deeper namespace, new file won't have it either. I would expect Resharper to do the same. 

But yes, it's annoying with VisualStudio.

**Reply by robloo** [reply]
You're the boss :) I would say that sounds like a bit of a reach. True, namespace defaults can catch you sometimes but its standard practice to modify these and also to review them in commits. I dont know of any project that follows namespace directory structure so precisely that no modification is ever necessary. The irony here is Avalonia already has this issue so it wouldn't be a new problem at all.

**Comment by maxkatz6** [comment]
Merging of some assemblies is definitely planned
https://github.com/AvaloniaUI/Avalonia/pull/5831

**Comment by grokys** [comment]
Personally I'd love to! I often get frustrated about the inconsistencies. However there are 2 problems:

1. Most common pushback on such ideas is that: it introduces noise to `git blame`, makes cherry-picking between master and release branches more difficult, and causes extra work for people with open PRs
2. If we don't have an automated tool to fix such errors before committing, we'll be back here in 2 years

It's a difficult balance between the camp who wants consistency and the camp who feels that consistency is a distraction and just want to get on with solving problems. I'm in camp A, btw.

**Reply by pr8x** [reply]
I think I said that before, but: Sacrificing readability, consistency and by that maintainability of a code base because it introduces minor inconveniences to some git tools is really stupid. We could have a one month notice period to flush pending/obsolete PRs (that should be done frequently anyway), so they won't conflict too much.

 `git blame` actually allows to ignore a certain commit via `--ignore-rev` (see https://akrabat.com/ignoring-revisions-with-git-blame/). We can even ignore an entire list of commits via `--ignore-revs-file` so whenever reformat patches are done we add the commit hash in there.

**Reply by robloo** [reply]
Fully agree with @pr8x and have also said elsewhere git blame is a poor reason to reject changes like this. 

However, I also agree that automated tooling is needed to enforce this going forward.

**Reply by maxkatz6** [reply]
In some cases it makes "git blame" completely broken without any chance to fix it (yes, I am looking at you, GitHub, who does not follow moved files in blame view).

**Reply by maxkatz6** [reply]
Although I still agree, that it makes sense anyway, especially when we already planning to break git history with https://github.com/AvaloniaUI/Avalonia/pull/5831

**Reply by robloo** [reply]
Yea, support for --follow in GitHub is a sore subject. They've been asked to add it almost since the beginning and there is no movement. I'm guessing it requires a lot more processing power or something and they just don't want their server costs to go up. That said, they don't need to run it by default and it could be an additional button in history like an old extension someone wrote for chrome.

**Comment by grokys** [comment]
Ok, so if we were to fix up formatting as part of https://github.com/AvaloniaUI/Avalonia/pull/5831 does anyone have suggestions for preventing the codebase getting into a mess again?

The options I'm aware of are Rider/Resharper and StyleCop. Anything else? Something like [`gofmt`](https://blog.golang.org/gofmt) for C#? Anyone used https://github.com/dotnet/format?

**Reply by workgroupengineering** [reply]
> The options I'm aware of are Rider/Resharper and StyleCop. Anything else? Something like [`gofmt`](https://blog.golang.org/gofmt) for C#? Anyone used https://github.com/dotnet/format?

Xamarin use it. They have a GitAction that daily check code formatting, see [here](https://github.com/xamarin/Xamarin.Forms/blob/5.0.0/.github/workflows/dotnet-format-daily.yml)

**Reply by robloo** [reply]
It looks like dotnet-format can also use existing analyzers and apply their fixes. The following analyzers seem best on my end: 

 * Microsoft.CodeAnalysis.NetAnalyzers
 * StyleCop.Analyzers

I think this should only be run on PRs for the most part as well. It seems like a bad idea to have a continuously running action that's attempting to modify source code in the entire solution.

Additionally, the Uno Platform uses Codacy to check code in each PR.

It also doesn't look like dotnet-format is powerful enough to organize source code itself. For example it can't order class files like:

 1. Fields
 2. Properties
 2. Constructor/Destructor
 3. Public then Private Methods

in whatever order is decided.

**Reply by kekekeks** [reply]
The list of requirements for a formatting tool:
1) support for VS and Rider and VSCode
2) format-on-save in at least VS and Rider and format-on-build for everything else
3) _no commit noise_
4) formatting is checked on CI

**Reply by robloo** [reply]
I'm been keeping this conversation in mind as I read things over the past few months. It looks like there are two tools that would work here:

1. The .NET Code Formatter Microsoft uses: https://github.com/dotnet/codeformatter. This looks like it was abandoned around 2018 and would likely need to be minimally supported again for usage.
2. JetBrains ReSharper : As JetBrains already uses Avalonia and also provides tooling to open source projects for free, I think they would work with the Avalonia Core team to set something up.

Avoiding 'commit noise' is impossible upon initial code formatting. However, it could be done for each PR after the initial code formatting. Cleaning up code to improve maintainability more than justifies a few extra commands to get git blame.

**Reply by kekekeks** [reply]
>However, it could be done for each PR after the initial code 

That would force us to squash PRs. We avoid doing that when possible.

**Reply by robloo** [reply]
A minor trade-off it seems to cleanup the code base. Many other repos automatically squash too. That has some benefits itself although you might disagree.

I've noticed the quality of code formatting with several recent commits in inconsistent and even suboptimal. The code base is going to get messy unless core maintainers become much more strict in code review or an automated tool is used. Everyone appreciates a clean code base that is easy to navigate and consistent.

**Reply by robloo** [reply]
I suppose there is a best-of-both worlds solution. Have a tool that analyzes PRs for code formatting errors and simply block merging until they are fixed. It would be a form of code analysis but basically raises errors on formatting differences. This would force committers to fix their formatting to follow the rules -- but it wouldn't be done automatically.

No tool exists to do this I'm aware of though tools like Codacy can do some of it.

**Reply by kekekeks** [reply]
Without reformat-on-build we are pretty much guaranteed to have "Fixed formatting" commits in PRs. I don't think there is an existing OSS solution for doing this, so I suppose that we need our own MSBuild task and either scrap CodeFormatter for parts or just use Roslyn directly since `Microsoft.CodeAnalysis.CSharp.Workspaces ` has a decent code formatter with configurable rule set.

We also need to decide if we want just code formatting or enforcement of file/class layout. Personally I don't like that because it doesn't allow to group logically interdependent parts like avalonia property declarations. There is also a problem with auto-fixing such things. We could use roslyn, but it might break stuff like conditional compilation. Without auto-fixing such rule will be _way too annoying_.

**Reply by robloo** [reply]
> MSBuild task and either scrap CodeFormatter for parts or just use Roslyn directly since Microsoft.CodeAnalysis.CSharp.Workspaces 

Yes, a Roslyn-based solution is likely the best open-source option. Though JetBrains internal tooling is probably [still more powerful and general purpose](https://blog.jetbrains.com/dotnet/2014/04/10/resharper-and-roslyn-qa/). Microsoft is struggling these days (less so after .NET Core) and isn't at the level of skill they used to be. At the turn of the century they had the best engineers and great management -- not so these days. All this is to say, like Avalonia itself, the less you rely on Microsoft the better.... I digress

> enforcement of file/class layout.

Personally, I would require this. Rules like: fields at the top of the file, organized members by public/private, methods all together, etc. I still like the idea of putting Properties in their own `<Control>.Properties.cs` file as WinUI is doing these days. It separates them out from logic and just makes things easier to read -- and alphabetically sorted. I know you guys don't want to do this yet though.

Finally, I think this should be opt-out with a comment analogous to `// <auto-generated />` which disables code analysis. The example you gave where some things make more sense to arrange by logically interdependent parts is a good special case. Also source code pulled from WPF/WinUI like Grid should not be changed to avoid loosing the ability to copy over updates. This would mitigate some of your other concerns as well.
